### PR TITLE
[MONO] Add JIT_CODE_DEBUG_INFO record functionality for Jitdump

### PR DESCRIPTION
**Pull request by saitama951**
This is related to #36774 . added support for `JIT_CODE_DEBUG_INFO` record which helps to annotate source code in perf.

Changed the jitdump version to 1 as the perf sources expect version 1

https://github.com/torvalds/linux/blob/master/tools/perf/util/jitdump.h#L25

Example use:-
    MONO_ENV_OPTIONS="--jitdump" perf record -k 1 dotnet <path/to/binary>
    perf inject --jit -i perf.data -o perf.jit.data
    perf report -i perf.jit.data

cc: @fanyang-mono

### CODE REVIEW

File: src/mono/mono/mini/mini-runtime.c [hunk 1]
```diff
@@ -2128,7 +2145,75 @@ mono_emit_jit_dump (MonoJitInfo *jinfo, gpointer code)
 
 		record.code_index = ++code_index;
 
-		// TODO: write debugInfo and unwindInfo immediately before the JitCodeLoadRecord (while lock is held).
+		DebugEntry ent;
```
Comment (vargaz) [inline src/mono/mono/mini/mini-runtime.c:2148]: This should follow the mono coding conventions, see the rest of the file for examples.

Comment (saitama951) [reply src/mono/mono/mini/mini-runtime.c:2148]: done. let me know if there is anything else

**Review by saitama951 — COMMENTED** [review]
changed the style to adhere to the mono coding conventions

File: src/mono/mono/mini/mini-runtime.c [hunk 2]
```diff
@@ -2062,7 +2064,22 @@ typedef struct
 	// Null terminated function name
 	// Native code
 } JitCodeLoadRecord;
+typedef struct
+{
+        guint64 code_addr;
+        guint32 line;
+        guint32 discrim;
+        char name[];
+}DebugEntry;
```
Comment (fanyang-mono) [inline src/mono/mono/mini/mini-runtime.c:2073]:
```suggestion
} DebugEntry;
```

File: src/mono/mono/mini/mini-runtime.c [hunk 3]
```diff
@@ -2062,7 +2064,22 @@ typedef struct
...
+typedef struct
+{
+        RecordHeader header;
+        guint64 code_addr;
+        guint64 nr_entry;
+        DebugEntry debug_entry[];
+}JitCodeDebug;
```
Comment (fanyang-mono) [inline src/mono/mono/mini/mini-runtime.c:2080]:
```suggestion
} JitCodeDebug;
```

File: src/mono/mono/mini/mini-runtime.c [hunk 4]
```diff
@@ -2062,7 +2064,22 @@ typedef struct
...
 
+static void add_basic_JitCodeDebug_info(JitCodeDebug *record);
```
Comment (fanyang-mono) [inline src/mono/mono/mini/mini-runtime.c:2082]:
```suggestion
static void add_basic_JitCodeDebug_info (JitCodeDebug *record);
```

Comment (fanyang-mono) [reply src/mono/mono/mini/mini-runtime.c:2082]: In mono code base, we add a space between the function name and the "(". Please follow this function call style for all the code change in this PR.

File: src/mono/mono/mini/mini-runtime.c [hunk 5]
```diff
@@ -2128,7 +2145,72 @@ mono_emit_jit_dump (MonoJitInfo *jinfo, gpointer code)
...
+		
+		//populating info relating debug methods
+		minfo = mono_debug_lookup_method(jinfo->d.method);
+                dmji = mono_debug_find_method( jinfo->d.method, NULL);
+
+                add_basic_JitCodeDebug_info(&rec);
+                rec.code_addr = (guint64)dmji->code_start;
+                rec.header.total_size = sizeof(rec) + sizeof(ent) + 1;
+                rec.nr_entry=1;
+                for(i=0;i < dmji->num_line_numbers;++i){
```
Comment (fanyang-mono) [inline src/mono/mono/mini/mini-runtime.c:2165]: Fix code style

File: src/mono/mono/mini/mini-runtime.c [hunk 6]
```diff
@@ -2128,7 +2145,72 @@ mono_emit_jit_dump (MonoJitInfo *jinfo, gpointer code)
...
+                        
+			loc = mono_debug_lookup_source_location_by_il(jinfo->d.method,dmji->line_numbers[i].il_offset,NULL);
+
+                        if(!(loc)
```
Comment (fanyang-mono) [inline src/mono/mono/mini/mini-runtime.c:2169]:
```suggestion
                        if(!loc)
```

**Review by fanyang-mono — COMMENTED** [review]
Please go through your change and fix the coding style. I've pointed out a few examples for you.

**Comment by saitama951** [comment]
> Please go through your change and fix the coding style. I've pointed out a few examples for you.

Thank you for helping me out with the code style. I have made the changes accordingly.

**Comment by saitama951** [comment]
@fanyang-mono is there any other changes required to be made? 
maybe we can push the code?

**Comment by directhex** [comment]
@vargaz @lambdageek can someone see if this is ready to merge now?

**Review by fanyang-mono — APPROVED** [review]

**Comment by shivanirmishra** [comment]
@fanyang-mono 
The fix did not work for me.
The steps we tried are:
```
1) dotnet build
2) MONO_ENV_OPTIONS="--jitmap --jitdump" perf record -k 1 dotnet run
3) perf inject --jit -i perf.data -o perf.jit.data
4) perf report -i perf.jit.data
```
.map and .dump files are getting generated in /tmp/ folder

Error:
`objdump: /tmp/perf-886285.map: File format not recognized`

Perf tool is able to read .map files to do symbol resolution and show the data. And due to this, we are able to see the functions in the profile but not the annotations inside that which requires a .dump file.

**Comment by saitama951** [comment]
@shivanirmishra The runtime should generate the dump file. can you cross check that under the `/tmp/`? 
if that is present can you check if the dumps are read via the strace?

**Comment by shivanirmishra** [comment]
Hi @saitama951, yes runtime is generating the .dump file in the/tmp/ folder.
But Strace is not showing any references to dump files.
The observations we had are with .NET7.0 so is the above fix available only on .NET8 or later?

**Comment by saitama951** [comment]
@shivanirmishra the fix should be there in .NET 8 preview 4 or you can cherry pick the commit.

**Comment by shivanirmishra** [comment]
@saitama951 
We tried with .NET8 Preview 4, but getting the below issue after using `perf inject`:
image
Also when we are doing Strace on the same, we can see the MMAP calls there:
image
So, what can cause this `wrong jitdump version 2, expected 1` issue in `perf inject`?

**Comment by saitama951** [comment]
@shivanirmishra I have fixed the `wrong versioning of jitdump` in the current PR itself, probably the fixes aren't incorporated in this release I guess , I would suggest you can build dotnet from source that should work fine.

**Comment by shivanirmishra** [comment]
Hi @saitama951, thank you for your help.
This fix is working for the sample dotnet code.
But while building dotnet for the workload from `https://github.com/dotnet/performance`, we are facing the below issues with the power system:
image
Any suggestion regarding the above error?
Also, I have changed the target frameworks to .net8 from .net7.

**Comment by shivanirmishra** [comment]
@saitama951 @fanyang-mono 
We are facing the above issue on .net8 on power, so will this backporting to 7.0 fix the issue on .net8 as well?

**Comment by saitama951** [comment]
@shivanirmishra It seems source code annotation fails on top of the BenchmarkDotnet library. it works for sample programs without the BDN. I have been manually looking into the jitted-so's, everything looks good there as well.

**Comment by shivanirmishra** [comment]
Thanks @saitama951 
It works for other workloads that run for a shorter duration.
But if we are trying to run a longer workload, then sometimes it's showing "unknown" in the profile on both X86 and Power systems.

For Example: after running the `BinaryTrees_5` workload from [https://github.com/dotnet/performance](https://github.com/dotnet/performance), we get the below profiles.

Before Inject:
image
After Inject:
image

**Comment by fanyang-mono** [comment]
> @saitama951 @fanyang-mono We are facing the above issue on .net8 on power, so will this backporting to 7.0 fix the issue on .net8 as well?

Yes, I am working on backporting the fixes to regressions to 7.0 now.

**Comment by fanyang-mono** [comment]
@shivanirmishra We will get this fix in the next possible servicing release.

**Comment by saitama951** [comment]
@shivanirmishra I have proposed the fix for the issue that you were facing.

https://github.com/dotnet/runtime/pull/88373

**Comment by shivanirmishra** [comment]
Hi @saitama951 @fanyang-mono 
This is another sample code for matrix multiplication, in which we are seeing the same above issue.

> MONO_ENV_OPTIONS="--jitmap --jitdump" perf record -k 1 /home/user/dotnet run

Dimension of the matrix: 2000x3 and 3x2000, and running the **MultiplyMatrices** function once.

> perf Report -i perf.data

image


> perf inject -j -i ./perf.data -o ./perf.jit.data

> perf report -i perf.jit.data

image

We are not able to see the main function **MultiplyMatrices** anywhere in the profile.
And wherever we are seeing the project name (new-matrix) as a command in the profile, we don’t see any functions/annotations for some so it might be splitting the main function into some parts and we are not able to resolve the same.

image

**Comment by shivanirmishra** [comment]
@saitama951 Yes, It's working.
Thank you.

**Pull request merged**